File: make_single_board.py
# Create a bingo board json from a txt
# Ex.
# python make_single_board.py --iterations 50000000 --tolerance 1 --showImage

import numpy as np
import sys
import os
import json
import argparse
import re
import time
from tqdm import tqdm
import matplotlib.pyplot as plt
import platform

# Parse arguments
parser = argparse.ArgumentParser(description="parameters")
parser.add_argument("--input", type=str, help="Path to tile list .txt", default="16star.txt")
parser.add_argument("--iterations", type=int, help="Max iterations to attempt", default=1000000)
parser.add_argument("--tolerance", type=float, help="Max allowed difference in difficulty between easiest and hardest row/col", default=1)
parser.add_argument("--seed", type=int, help="Random seed")
parser.add_argument("--showImage", action='store_true', help="Show debug image of final board")
args = parser.parse_args()
bingo_file = args.input
iterations = args.iterations
tolerance = args.tolerance
showImage = args.showImage
seed = args.seed if args.seed is not None else int(time.time())

# Check if file exists
if not os.path.exists(bingo_file):
    bingo_file = "data/" + bingo_file
if not os.path.exists(bingo_file):
    print("Error: file not found:", bingo_file)
    exit(1)

# Convert board to json format
def board_to_json(board):
    result = []
    for i in range(25):
        result.append({ 'name': board[i]['name'], 'score': board[i]['score'] })

    return json.dumps(result)

def is_magic_square_with_tolerance(arr, tolerance):
    row_sums = arr.sum(axis=1)
    col_sums = arr.sum(axis=0)
    main_diag_sum = np.sum(np.diagonal(arr))
    anti_diag_sum = np.sum(np.diagonal(np.fliplr(arr)))
    
    # Combine sums into a single array
    all_sums = np.concatenate((row_sums, col_sums, [main_diag_sum, anti_diag_sum]))
    return (np.max(all_sums) - np.min(all_sums)) <= tolerance

# Display a debug image showing grid with tile names
def show_board_preview(names):
    # Reshape the list into a 5x5 grid (2D array)
    names_grid = np.array(names).reshape(5, 5)

    # Create a plot with no axes
    fig, ax = plt.subplots(figsize=(6, 6))
    ax.axis('off')

    # Display each name in the grid
    for i in range(5):
        for j in range(5):
            ax.text(j, 4-i, names_grid[i, j], ha='center', va='center', fontsize=5)

    # Add gridlines between cells
    ax.set_xticks(np.arange(0, 6, 1))  # X grid lines
    ax.set_yticks(np.arange(0, 6, 1))  # Y grid lines
    ax.grid(which='both', color='black', linestyle='-', linewidth=1)  # Customize gridline appearance

    # Show the plot
    plt.show()

def clear_console():
    if platform.system() == "Windows":
        os.system("cls")
    else:  # For macOS and Linux
        os.system("clear")

def print_success():
    print(" _       __    _                   __                    _ __                       _       __")
    print("| |     / /__ ( )   _____     ____/ /___  ____  ___     (_) /_   ____ _____ _____ _(_)___  / /")
    print("| | /| / / _ \|/ | / / _ \   / __  / __ \/ __ \/ _ \   / / __/  / __ `/ __ `/ __ `/ / __ \/ / ")
    print("| |/ |/ /  __/ | |/ /  __/  / /_/ / /_/ / / / /  __/  / / /_   / /_/ / /_/ / /_/ / / / / /_/  ")
    print("|__/|__/\___/  |___/\___/   \__,_/\____/_/ /_/\___/  /_/\__/   \__,_/\__, /\__,_/_/_/ /_(_)   ")
    print("                                                                    /____/                    ")

# List of every tile name with difficulty score
lines = np.genfromtxt(bingo_file.strip(), dtype='str', delimiter='/n')

# Get all difficulty scores
# Score is the value between "[]"
difficulty_scores = []
for line in lines:
    match = re.search(r"\[([\d.]+)\]", line)        # find value between "[]"
    difficulty_scores.append(float(match.group(1))) # extract value between "[]" and append to list

# Shuffle tiles
# Break if conditions are met for a good board
difficulty_scores = np.array(difficulty_scores)
lenVal = len(difficulty_scores)
t0 = time.time()
success = False
for _ in tqdm(range(iterations), desc="Crunching the Numbers"):

    # Get random order for candidate board
    shuffled_indices = np.random.choice(lenVal, size=25, replace=False)

    # Create random board
    scores_grid = difficulty_scores[shuffled_indices].reshape(5, 5)

    # Check that the sums of each row/col pass conditions
    if is_magic_square_with_tolerance(scores_grid, tolerance=tolerance):
        success = True
        break

# Print results
clear_console()
if success:
    print_success()
    row_sums = np.sum(scores_grid, axis=1)
    col_sums = np.sum(scores_grid, axis=0)
    main_diag_sum = np.sum(np.diagonal(scores_grid))
    anti_diag_sum = np.sum(np.diagonal(np.fliplr(scores_grid)))
    all_sums = np.concatenate([row_sums, col_sums, [main_diag_sum, anti_diag_sum]])
    max_sum = np.max(all_sums)
    min_sum = np.min(all_sums)
    print("\nHardest Difficulty Row Sum: ", max_sum)
    print("Easiest Difficulty Row Sum: ", min_sum)

    # Print final scores grid
    print("\nFinal GRID:")
    print(scores_grid)

    # Create json
    out = []
    names_list = []
    cnt = 0
    scores_rand = difficulty_scores[shuffled_indices]
    for i, score in enumerate(scores_rand):
        name = lines[shuffled_indices[i]]
        out.append({ 'name': name, 'score': score })
        names_list.append(name)
        cnt += 1
        if cnt == 25:
            break
    out = np.array(out)
    print("\nFinal JSON:")
    print(board_to_json(out))

else:
    print("Failed to make board")

# Print time to run
print("\nTook ", np.round(time.time()-t0,2), " sec")

# Show debug image
if showImage:
    show_board_preview(names_list)

File: make_single_board.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <math.h>
#include <stdbool.h>

#define MAX_LINES 1000
#define LINE_LENGTH 256
#define BOARD_SIZE 25

// Structure to represent a bingo tile
typedef struct {
    char name[LINE_LENGTH];
    float score;
} Tile;

// Configuration parameters with default values
typedef struct {
    char input_file[256];
    int iterations;
    float tolerance;
    int seed;
    float minDifficulty;
    bool show_image;
} Config;

// Global variables
Tile tiles[MAX_LINES];
int total_tiles = 0;
Config config = {
    .input_file = "./data/16star.txt",
    .iterations = 100000000,
    .tolerance = 1.0,
    .seed = 0,
    .minDifficulty = 0.0,
    .show_image = false
};

// Function prototypes
void parse_arguments(int argc, char *argv[]);
void read_bingo_file();
bool is_magic_square_with_tolerance(float grid[5][5], float tolerance, float minDifficulty);
void generate_board();
void print_success();
void print_board_details(float grid[5][5]);

// Argument parsing
void parse_arguments(int argc, char *argv[]) {
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--input") == 0 && i + 1 < argc) {
            strcpy(config.input_file, argv[++i]);
        } else if (strcmp(argv[i], "--iterations") == 0 && i + 1 < argc) {
            config.iterations = atoi(argv[++i]);
        } else if (strcmp(argv[i], "--tolerance") == 0 && i + 1 < argc) {
            config.tolerance = atof(argv[++i]);
        } else if (strcmp(argv[i], "--seed") == 0 && i + 1 < argc) {
            config.seed = atoi(argv[++i]);
        } else if (strcmp(argv[i], "--showImage") == 0) {
            config.show_image = true;
        } else if (strcmp(argv[i], "--minDifficulty") == 0 && i + 1 < argc) {
            config.minDifficulty = atof(argv[++i]);
        }
    }
}

// Read bingo file and parse tiles
void read_bingo_file() {
    FILE *file = fopen(config.input_file, "r");
    if (file == NULL) {
        fprintf(stderr, "Error: Cannot open file %s\n", config.input_file);
        exit(1);
    }

    char line[LINE_LENGTH];
    while (fgets(line, sizeof(line), file) && total_tiles < MAX_LINES) {
        // Extract name and score from line
        char *score_start = strchr(line, '[');
        char *score_end = strchr(line, ']');
        
        if (score_start && score_end) {
            *score_end = '\0';
            tiles[total_tiles].score = atof(score_start + 1);
            *score_start = '\0';
            strcpy(tiles[total_tiles].name, line);
            total_tiles++;
        }
    }
    fclose(file);
}

// Check if grid meets magic square tolerance
// Also check if min score is above minDifficulty
bool is_magic_square_with_tolerance(float grid[5][5], float tolerance, float minVal) {
    float row_sums[5] = {0};
    float col_sums[5] = {0};
    float main_diag_sum = 0;
    float anti_diag_sum = 0;

    // Calculate row and column sums
    for (int i = 0; i < 5; i++) {
        for (int j = 0; j < 5; j++) {
            row_sums[i] += grid[i][j];
            col_sums[j] += grid[i][j];
        }
        main_diag_sum += grid[i][i];
        anti_diag_sum += grid[i][4-i];
    }

    // Find max and min sums
    float max_sum = row_sums[0];
    float min_sum = row_sums[0];
    
    for (int i = 0; i < 5; i++) {
        max_sum = fmaxf(max_sum, row_sums[i]);
        min_sum = fminf(min_sum, row_sums[i]);
        max_sum = fmaxf(max_sum, col_sums[i]);
        min_sum = fminf(min_sum, col_sums[i]);
    }
    max_sum = fmaxf(max_sum, main_diag_sum);
    max_sum = fmaxf(max_sum, anti_diag_sum);
    min_sum = fminf(min_sum, main_diag_sum);
    min_sum = fminf(min_sum, anti_diag_sum);

    if (min_sum < minVal) {
        return false;
    }

    return (max_sum - min_sum) <= tolerance;
}

// Generate bingo board
void generate_board() {
    // Set random seed
    srand(config.seed ? config.seed : time(NULL));

    bool success = false;
    float scores_grid[5][5];
    int shuffled_indices[25];

    // Loop through iterations until a sufficient board is found
    for (int iter = 0; iter < config.iterations; iter++) {
        // Create unique random selection of 25 tiles
        for (int i = 0; i < 25; i++) {
            int index;
            bool unique;
            do {
                unique = true;
                index = rand() % total_tiles;
                for (int j = 0; j < i; j++) {
                    if (shuffled_indices[j] == index) {
                        unique = false;
                        break;
                    }
                }
            } while (!unique);
            shuffled_indices[i] = index;
        }

        // Create grid of scores
        for (int i = 0; i < 25; i++) {
            int row = i / 5;
            int col = i % 5;
            scores_grid[row][col] = tiles[shuffled_indices[i]].score;
        }

        // Check magic square conditions
        if (is_magic_square_with_tolerance(scores_grid, config.tolerance, config.minDifficulty)) {
            success = true;
            
            // Print success and board details
            print_success();
            print_board_details(scores_grid);

            // Print JSON-like output
            printf("\nFinal JSON:\n[");
            for (int i = 0; i < 25; i++) {
                Tile tile = tiles[shuffled_indices[i]];
                printf("%s{\"name\": \"%s\", \"score\": %.2f}", 
                       i > 0 ? ", " : "", tile.name, tile.score);
            }
            printf("]\n");
            break;
        }
    }

    if (!success) {
        printf("Failed to make board\n");
    }
}

// Print ASCII success art
void print_success() {
    printf(" _       __    _                   __                    _ __                       _       __\n");
    printf("| |     / /__ ( )   _____     ____/ /___  ____  ___     (_) /_   ____ _____ _____ _(_)___  / /\n");
    printf("| | /| / / _ \\|/ | / / _ \\   / __  / __ \\/ __ \\/ _ \\   / / __/  / __ `/ __ `/ __ `/ / __ \\/ / \n");
    printf("| |/ |/ /  __/ | |/ /  __/  / /_/ / /_/ / / / /  __/  / / /_   / /_/ / /_/ / /_/ / / / / /_/  \n");
    printf("|__/|__/\\___/  |___/\\___/   \\__,_/\\____/_/ /_/\\___/  /_/\\__/   \\__,_/\\__, /\\__,_/_/_/ /_(_)   \n");
    printf("                                                                    /____/                    \n");
}

// Print board details
void print_board_details(float grid[5][5]) {
    float row_sums[5] = {0};
    float col_sums[5] = {0};
    float main_diag_sum = 0;
    float anti_diag_sum = 0;

    // Calculate sums
    for (int i = 0; i < 5; i++) {
        for (int j = 0; j < 5; j++) {
            row_sums[i] += grid[i][j];
            col_sums[j] += grid[i][j];
        }
        main_diag_sum += grid[i][i];
        anti_diag_sum += grid[i][4-i];
    }

    // Get min/max sums for printing
    float min_sum = INFINITY;
    float max_sum = -INFINITY;

    // Check row sums
    for (int i = 0; i < 5; i++) {
        min_sum = fminf(min_sum, row_sums[i]);
        max_sum = fmaxf(max_sum, row_sums[i]);
    }

    // Check column sums
    for (int j = 0; j < 5; j++) {
        min_sum = fminf(min_sum, col_sums[j]);
        max_sum = fmaxf(max_sum, col_sums[j]);
    }

    // Check diagonal sums
    min_sum = fminf(min_sum, main_diag_sum);
    max_sum = fmaxf(max_sum, main_diag_sum);
    min_sum = fminf(min_sum, anti_diag_sum);
    max_sum = fmaxf(max_sum, anti_diag_sum);

    printf("\nHardest Sum: %.1f\n", max_sum);
    printf("Easiest Sum: %.1f\n", min_sum);

    printf("\nFinal GRID:\n");
    for (int i = 0; i < 5; i++) {
        for (int j = 0; j < 5; j++) {
            printf("%.2f ", grid[i][j]);
        }
        printf("\n");
    }
}

int main(int argc, char *argv[]) {
    // Parse command line arguments
    parse_arguments(argc, argv);

    // Read bingo tile file
    read_bingo_file();

    // Generate board
    generate_board();

    return 0;
}